include/boost/gil/image_processing/threshold.hpp

```diff
@@ -334,6 +343,115 @@ void threshold_optimal
     }
 }
 
+namespace detail {
+
+template
+<
+    typename SourceChannelT,
+    typename ResultChannelT,
+    typename SrcView,
+    typename DstView,
+    typename Operator
+>
+void adaptive_impl
+(
+    SrcView const& src_view,
+    SrcView const& convolved_view,
+    DstView const& dst_view,
+    Operator const& threshold_op
+)
+{
+    //template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+
+    static_assert(color_spaces_are_compatible
+    <
+        typename color_space_type<SrcView>::type,
+        typename color_space_type<DstView>::type
+    >::value, "Source and destination views must have pixels with the same color space");
+
+    //iterate over the image chaecking each pixel value for the threshold
+    for (std::ptrdiff_t y = 0; y < src_view.height(); y++)
+    {
+        typename SrcView::x_iterator src_it = src_view.row_begin(y);
+        typename SrcView::x_iterator convolved_it = convolved_view.row_begin(y);
+        typename DstView::x_iterator dst_it = dst_view.row_begin(y);
+
+        for (std::ptrdiff_t x = 0; x < src_view.width(); x++)
+        {
+            static_transform(src_it[x], convolved_it[x], dst_it[x], threshold_op);
+        }
+    }
+}
+} //namespace boost::gil::detail
+
+template <typename SrcView, typename DstView>
+void threshold_adaptive
+(
+    SrcView const& src_view,
+    DstView const& dst_view,
+    typename channel_type<DstView>::type max_value,
+    std::size_t kernel_size,
+    threshold_adaptive_method method = threshold_adaptive_method::mean,
+    threshold_direction direction = threshold_direction::regular
+)
+{
+    BOOST_ASSERT_MSG((kernel_size % 2 != 0), "Kernel size must be an odd number");
+
+    typedef typename channel_type<SrcView>::type source_channel_t;
+    typedef typename channel_type<DstView>::type result_channel_t;
+
+    if (method == threshold_adaptive_method::mean)
+    {
+        float *mean = new float[kernel_size];
+        std::fill_n(mean, kernel_size, 1.0f/kernel_size);
+
+        image<typename SrcView::value_type> temp_img(src_view.width(), src_view.height());
+        typename image<typename SrcView::value_type>::view_t temp_view = view(temp_img);
+        SrcView temp_conv(temp_view);
+        
+        kernel_1d<float> kernel(mean, kernel_size, kernel_size/2);
+        convolve_rows<pixel<float, typename SrcView::value_type::layout_t>>(
+            src_view, kernel, temp_view
+        );
+        convolve_cols<pixel<float, typename SrcView::value_type::layout_t>>(
+            temp_view, kernel, temp_view
+        );
+
+        if (direction == threshold_direction::regular)
+        {
+            detail::adaptive_impl<source_channel_t, result_channel_t>(src_view, temp_conv, dst_view,
+                [max_value](source_channel_t px1, source_channel_t px2) -> result_channel_t
+                    { return px1 >= px2 ? max_value : 0; });
```

Comment:
This is a regular binary threshold, so shouldn't `px1 >= px2` read `px1 > px2`?

Could you also rename
- `px1` to just `px` or `pixel`  
- `px2` to `threshold`
?

While reading this, I constantly have to double-check in memory which is which, which comes  from source pixels and which from convolved pixels as thresholds :-)

Reply:
From [your yesterday comment on Gitter](https://gitter.im/boostorg/gil?at=5d379c0bb2db751dabb98df8):
> - correct output is what OpenCV produces with `C=0`
> - correct output is **white rectangles on black background**

I don't think there is a single correct output because for such sharp edge cases (i.e. dark gray features on light gray background), thresholding result does depend on value of `C1:

From adaptive threshold description at http://homepages.inf.ed.ac.uk/rbf/HIPR2/adpthrsh.htm
> if the threshold employed is not the mean, but (mean-C), where C is a constant.
> Using this statistic, all pixels which exist in a uniform neighborhood (e.g. along the margins) are set to background.

Assuming that [OpenCV defaults to C=5](https://github.com/opencv/opencv/blob/199ddff13b73071529a42cff97ee8286cfc15d02/modules/imgproc/include/opencv2/imgproc/imgproc_c.h#L833), one may argue that 'correct' output is **black rectangles on white background**. This is what OpenCV produces from our `3000x3000` test image with the two rectangles for any `C>0`. So, since 'correct' depends on algorithm parameters, I wouldn't seek for a single correct answer.

OpenCV:

- `C > 0`: **black rectangles** on white background
- `C = 0`: **white rectangles** on black background

GIL:

- `px1 >= px2 ? max_value : 0`: **black rectangles** on white background
- `px1 >  px2 ? max_value : 0`: **white rectangles** on black background

In other words, equivalent threshold tests are:

* **black rectangles** 

    ```
    px1 >= px2      ? max_value : 0
    px1 > (px2 - 5) ? max_value : 0
    ```

* **white rectangles**

    ```
    px1 > px2       ? max_value : 0
    px1 > (px2 - 0) ? max_value : 0
    ```

So, the major difference between behaviour of OpenCV and GIL that I can see is this flexibility due to the extra constant parameter. Your algorithm could also allow such parametrization.

Reply:
Additional notes on the point 1. above posted in https://github.com/boostorg/gil/pull/341#issuecomment-514817002